File: python-sdk/expansion/interfaces/rpc/blueprints_library.py
from typing import Optional, List
from enum import Enum

import expansion.api as api
from expansion.transport import IOTerminal, Channel
from expansion import utils
from expansion.types import Blueprint


class Status(Enum):
    SUCCESS = "success"
    INTERNAL_ERROR = "internal error"
    BLUEPRINT_NOT_FOUND = "blueprint not found"

    # Problems, detected on SDK side
    FAILED_TO_SEND_REQUEST = "failed to send request"
    UNREACHABLE = "remote side is unreachable"
    RESPONSE_TIMEOUT = "response timeout"
    UNEXPECTED_RESPONSE = "unexpected response"
    CHANNEL_CLOSED = "channel closed"
    CANCELED = "operation canceled"

    def is_success(self):
        return self == Status.SUCCESS

    @staticmethod
    def from_protobuf(status: api.IResourceContainer.Status) -> "Status":
        PbStatus = api.IBlueprintsLibrary.Status
        return {
            PbStatus.SUCCESS: Status.SUCCESS,
            PbStatus.INTERNAL_ERROR: Status.INTERNAL_ERROR,
            PbStatus.BLUEPRINT_NOT_FOUND: Status.BLUEPRINT_NOT_FOUND,
        }[status]


class BlueprintsLibraryI(IOTerminal):

    def __init__(self, name: Optional[str] = None):
        if not name:
            name = utils.generate_name(BlueprintsLibraryI)
        super().__init__(name=name)

    @Channel.return_on_close(Status.CHANNEL_CLOSED, [])
    async def get_blueprints_list(self,
                                  start_with: str = "",
                                  timeout: float = 0.5) -> (Status, List[str]):
        request = api.Message()
        request.blueprints_library.blueprints_list_req = start_with
        if not self.send(request):
            return Status.FAILED_TO_SEND_REQUEST, []

        names: List[str] = []
        done = False
        while not done:
            response, _ = await self.wait_message(timeout=timeout)
            if not response:
                return Status.RESPONSE_TIMEOUT, []
            response = api.get_message_field(
                response,
                ["blueprints_library", "blueprints_list"])
            if not response:
                return Status.UNEXPECTED_RESPONSE, []
            names.extend(response.names)
            done = response.left == 0
        return Status.SUCCESS, names

    @Channel.return_on_close(Status.CHANNEL_CLOSED, None)
    async def get_blueprint(self,
                            blueprint_name: str,
                            timeout: float = 0.5) \
            -> (Status, Optional[Blueprint]):
        request = api.Message()
        request.blueprints_library.blueprint_req = blueprint_name
        if not self.send(request):
            return Status.FAILED_TO_SEND_REQUEST, None
        response, _ = await self.wait_message(timeout=timeout)
        if not response:
            return Status.RESPONSE_TIMEOUT, None

        blueprint = api.get_message_field(
                response, ["blueprints_library", "blueprint"])
        if blueprint:
            return Status.SUCCESS, Blueprint.from_protobuf(blueprint)

        fail = api.get_message_field(
            response, ["blueprints_library", "blueprint_fail"])
        if fail:
            return Status.from_protobuf(fail), None

File: typescript-sdk/lowlevel/blueprints_library.ts
import { create } from "@bufbuild/protobuf";
import * as msg from "../Protocol_pb.js";
import * as types from "../types/index.js";
import { Session } from "./session.js";


export type BlueprintsLibraryStatus =
    | "SUCCESS"
    | "INTERNAL_ERROR"
    | "BLUEPRINT_NOT_FOUND";

export type BlueprintsNamesPage = {
    names: string[];
    left: number;
}

export type BlueprintResult =
    | { case: "blueprint"; blueprint: types.Blueprint }
    | { case: "blueprint_fail"; status: BlueprintsLibraryStatus };

export class BlueprintsLibrary {

    constructor(private session: Session) {}

    async send_blueprints_list_request(start_with: string = ""): Promise<types.Status> {
        const request = create(msg.IBlueprintsLibrarySchema, {
            choice: { case: "blueprintsListReq", value: start_with },
        });
        return this.send(request);
    }

    async wait_blueprints_list(timeout: number = 500)
    : Promise<[types.Status, BlueprintsNamesPage | undefined]>
    {
        const [status, response] = await this.wait(timeout);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "blueprintsList") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        const page = response.choice.value;
        return [types.Status.ok(), {
            names: page.names ?? [],
            left: page.left,
        }];
    }

    async send_blueprint_request(blueprint_name: string): Promise<types.Status> {
        const request = create(msg.IBlueprintsLibrarySchema, {
            choice: { case: "blueprintReq", value: blueprint_name },
        });
        return this.send(request);
    }

    async wait_blueprint(timeout: number = 500)
    : Promise<[types.Status, BlueprintResult | undefined]>
    {
        const [status, response] = await this.wait(timeout);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case == "blueprintFail") {
            const server_status = blueprintsLibraryStatusFromProtobuf(response.choice.value);
            if (!server_status) {
                return [types.Status.fail(`got unexpected status ${response.choice.value}`),
                        undefined];
            }
            return [types.Status.ok(), {
                case: "blueprint_fail",
                status: server_status,
            }];
        }
        if (response.choice.case != "blueprint") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        return [types.Status.ok(), {
            case: "blueprint",
            blueprint: types.blueprintFromProtobuf(response.choice.value),
        }];
    }

    private async send(request: msg.IBlueprintsLibrary): Promise<types.Status> {
        const message = create(msg.MessageSchema, {
            choice: { case: "blueprintsLibrary", value: request },
        });
        return this.session.send(message);
    }

    private async wait(timeout_ms: number = 500)
    : Promise<[types.Status, msg.IBlueprintsLibrary | undefined]>
    {
        const [status, response] = await this.session.wait(timeout_ms);
        if (!status.is_ok() || !response) {
            return [status.wrap("no response"), undefined];
        }
        if (response.choice.case != "blueprintsLibrary") {
            return [types.Status.fail(`got unexpected message ${response.choice.case}`),
                    undefined];
        }
        return [types.Status.ok(), response.choice.value];
    }

}

function blueprintsLibraryStatusFromProtobuf(
    value: msg.IBlueprintsLibrary_Status): BlueprintsLibraryStatus | undefined
{
    switch (value) {
        case msg.IBlueprintsLibrary_Status.SUCCESS: return "SUCCESS";
        case msg.IBlueprintsLibrary_Status.INTERNAL_ERROR: return "INTERNAL_ERROR";
        case msg.IBlueprintsLibrary_Status.BLUEPRINT_NOT_FOUND: return "BLUEPRINT_NOT_FOUND";
        default: return undefined;
    }
}
